### PR TITLE
Add uTP support (BEP29)

### PR DESCRIPTION
**Pull request by hicom150**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix
[x] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
Heavily inspired in #1328 and #1486 this PR adds uTP (BEP29) support implemented over current master branch and with all tests passing.

fixes https://github.com/webtorrent/webtorrent/issues/68
closes #1486

**Which issue (if any) does this pull request address?**

**Is there anything you'd like reviewers to focus on?**
I initialized uTP as false by default until we find it is stable and works as expected 😅
I've also added some test that are quite slow as they wait for timeouts, so we can disable them once everything works as expected 😉

### CODE REVIEW

File: lib/conn-pool.js [hunk 1]
```diff
@@ -92,7 +127,7 @@ class TCPPool {
     self._pendingConns.push(conn)
     conn.once('close', cleanupPending)
 
-    const peer = Peer.createTCPIncomingPeer(conn)
+    const peer = this.utpServer ? Peer.createUTPIncomingPeer(conn) : Peer.createTCPIncomingPeer(conn)
```
Comment (KayleePop) [inline lib/conn-pool.js:130]: Doesn't this assume that if UTP is enabled every incoming connection will be UtP? (from both TCPServer and UTPServer)

File: lib/torrent.js [hunk 2]
```diff
@@ -750,7 +751,7 @@ class Torrent extends EventEmitter {
       }
     }
 
-    const wasAdded = !!this._addPeer(peer)
+    const wasAdded = !!this._addPeer(peer, this.client.utp ? 'utp' : 'tcp')
```
Comment (KayleePop) [inline lib/torrent.js:754]: just a note for other reviewers: if the utp connection fails to connect, then it is replaced with a tcp connection to the same ip:port

maybe a comment here explaining that would be good

File: lib/peer.js [hunk 3]
```diff
@@ -45,6 +46,22 @@ exports.createTCPIncomingPeer = conn => {
...
+  const addr = `${conn.remoteAddress}:${conn.remotePort}`
+  const peer = new Peer(addr, 'utpIncoming')
+  peer.conn = conn
+  peer.addr = addr
+
+  peer.onConnect()
+
+  return peer
+}
+
```
Comment (KayleePop) [inline lib/peer.js:64]: seems like you could make these into one function to be more DRY but not important.

File: lib/peer.js [hunk 4]
```diff
@@ -57,6 +74,18 @@ exports.createTCPOutgoingPeer = (addr, swarm) => {
...
+ * available connection), the client can attempt to connect to the address.
+ */
+exports.createUTPOutgoingPeer = (addr, swarm) => {
+  const peer = new Peer(addr, 'utpOutgoing')
+  peer.addr = addr
+  peer.swarm = swarm
+
+  return peer
+}
+
```
Comment (KayleePop) [inline lib/peer.js:88]: seems like you could make these into one function to be more DRY but not important.

**Review by KayleePop — CHANGES_REQUESTED** [review]
the tests don't actually make sure UtP works, just that the downloads complete when the `utp` setting is used on one or both clients. Not required to merge IMO, but it would be nice if possible to actually make sure UtP is working with CI

**Comment by KayleePop** [comment]
it would be good to give commit credit to @pldubouilh somehow for this too if that's possible

**Comment by pldubouilh** [comment]
Thanks ! Nice to revive these stalled commits :+1: 
One point to maybe keep in mind is that depending on `utp-native` would make webtorrent depend on a non-js, native binding module (afaiu for the first time).

File: lib/torrent.js [hunk 5]
```diff
@@ -24,6 +24,7 @@ const sha1 = require('simple-sha1')
 const speedometer = require('speedometer')
 const utMetadata = require('ut_metadata')
 const utPex = require('ut_pex') // browser exclude
+const utp = require('utp-native') // browser exclude
```
[the remark this answers is not in the document]
Comment (hicom150) [reply lib/torrent.js:27]: You are right! Done 👍

File: index.js [hunk 6]
```diff
@@ -14,7 +14,7 @@ const Peer = require('simple-peer')
 const randombytes = require('randombytes')
 const speedometer = require('speedometer')
 
-const TCPPool = require('./lib/tcp-pool') // browser exclude
+const ConnPool = require('./lib/conn-pool') // browser exclude
```
[the remark this answers is not in the document]
Comment (hicom150) [reply index.js:17]: Done 👍

File: lib/conn-pool.js (hunk 1 above)
Comment (hicom150) [reply lib/conn-pool.js]: As you pointed out it can cause confusion, so I've refactored a bit this part 😉

File: lib/torrent.js (hunk 2 above)
Comment (hicom150) [reply lib/torrent.js:755]: Right! I've added a comment 👍

File: lib/peer.js (hunk 3 above)
Comment (hicom150) [reply lib/peer.js:50]: Added a helper function to improve this 👍

File: lib/peer.js (hunk 4 above)
Comment (hicom150) [reply lib/peer.js]: Added a helper function to improve this 👍

**Review by KayleePop — APPROVED** [review]
:+1:

**Comment by hicom150** [comment]
@KayleePop First of all thank you for the feedback! 💪
I've made all the improvements you have suggested hope it is better now 😅
Regarding the tests that don't actually make sure UtP works, any suggestion on how to approach that would be more than welcome 😉
Last but not least, you are absolutely right that this is all based in the awesome PRs made by @pldubouilh so kudos to him! 🤟

**Comment by KayleePop** [comment]
you could force push with this https://github.blog/2018-01-29-commit-together-with-co-authors/ to give github credit to @pldubouilh 

not sure if that matters to them. Otherwise LGTM! uTP is a big win :rocket:

**Comment by hicom150** [comment]
I've been giving a though to this... 
> One point to maybe keep in mind is that depending on utp-native would make webtorrent depend on a non-js, native binding module (afaiu for the first time).

@pldubouilh
Do you think it could be more appropriate to use  https://github.com/mafintosh/utp-wasm instead of https://github.com/mafintosh/utp-native ? It seems that the share the same API although the wasm one is noted as "much slower" that the native one 😓

**Comment by feross** [comment]
Perhaps we can switch to `utp-wasm` if/when that library's performance improves. In the meantime, hopefully it's fine to use a native dependency -- we'll need to make sure the correct binary is used on each platform in WebTorrent Desktop.

If we end up having install errors on various platforms, then we can change `utp-native` to an optionalDependency and make sure that the library works even when it fails to install.

**Pull request merged**